### ISSUE TITLE
Add a modifier for "closed" functions

### ISSUE THREAD

**Issue by FilipeBeck** [microsoft/TypeScript#37028]
This issue is more restrictive than #7770. It prohibits access to any value outside the scope of the function.

Allowed accesses:
- arguments
- `this`, when is not a _arrow function_
- type declarations

Example of usage: A node function that creates a `Worker` by serializing a function provided as argument and using `eval` as an option for the constructor. Any value originally accessed that is outside the scope of the provided function will be `undefined` in the scope of the ` Worker`

**Comment by DanielRosenwasser** [comment]
So is the intent just to ensure that no outer variables are captured by the function?

**Comment by FilipeBeck** [comment]
Yes, that's it. Also, an compiler error should occur if a non-closed function is provided where a closed function is expected
